llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp

```diff
@@ -218,6 +218,17 @@ Instruction *InstCombinerImpl::visitMul(BinaryOperator &I) {
                   : BinaryOperator::CreateNeg(Op0);
   }
 
+  {
+    // abs(a * abs(b)) --> abs(a * b)
+    Value *A, *B;
+    if (match(&I, m_OneUse(m_c_Mul(m_Value(A),
+                                   m_Intrinsic<Intrinsic::abs>(m_Value(B)))))) {
+      Value *New =
+          HasNSW ? Builder.CreateNSWMul(A, B) : Builder.CreateMul(A, B);
```

Comment:
The nsw flag preserving is incorrect.
Alive2: https://alive2.llvm.org/ce/z/kXQA3p

Reply:
@dtcxzyw  I will update the patch now, and is the way that I used ` thwart complexity-based canonicalization` correct?